### ISSUE TITLE
`unused_mut` false positive due to borrow-checking error

### ISSUE THREAD

**Issue by theemathas** [rust-lang/rust#152024]
### Code

```Rust
struct Thing;
impl Drop for Thing {
    fn drop(&mut self) {}
}

fn main() {
    let mut t;
    let mut b = None;
    loop {
        t = Thing;
        b.insert(&t);
    }
}
```

### Current output

```Shell
warning: variable does not need to be mutable
 --> src/main.rs:7:9
  |
7 |     let mut t;
  |         ----^
  |         |
  |         help: remove this `mut`
  |
  = note: `#[warn(unused_mut)]` (part of `#[warn(unused)]`) on by default

error[E0506]: cannot assign to `t` because it is borrowed
  --> src/main.rs:10:9
   |
10 |         t = Thing;
   |         ^ `t` is assigned to here but it was already borrowed
11 |         b.insert(&t);
   |         -        -- `t` is borrowed here
   |         |
   |         borrow later used here

For more information about this error, try `rustc --explain E0506`.
warning: `playground` (bin "playground") generated 1 warning
error: could not compile `playground` (bin "playground") due to 1 previous error; 1 warning emitted
```

### Desired output

```Shell
The unused_mut lint should not fire
```

### Other cases

```Rust

```

### Rust Version

```Shell
Reproducible on the playground with version 1.95.0-nightly (2026-02-01 57d2fb136650d05efb3e)
```
